### ISSUE TITLE
Assertion when connecting to client running as root

### ISSUE THREAD

**Issue by occivink** [wolfpld/tracy#114]
Using tracy 0.7.2 on linux. More specifically, opensuse Tumbleweed.

I wanted to test out tracy, starting off with call stack sampling. After integrating tracy as client, running the client unprivileged works normally. When running the client as root, the server trips on the following assertion:
```
Tracy-debug: ../../../server/TracyWorker.cpp:3531: tracy::CallstackFrameId tracy::Worker::PackPointer(uint64_t) const: Assertion `( ( ptr & 0x3000000000000000 ) << 2 ) == ( ptr & 0xC000000000000000 )' failed.
```

This also occurs with a trivial client such as
```
#define TRACY_ENABLE

#include "TracyClient.cpp"

int main() { while (true) {} }
```

This is the stacktrace at the time of the crash:
```
#0  0x00007ffff63724b1 in raise () at /lib64/libc.so.6
#1  0x00007ffff635b539 in abort () at /lib64/libc.so.6
#2  0x00007ffff635b411 in _nl_load_domain.cold () at /lib64/libc.so.6
#3  0x00007ffff636a9e2 in  () at /lib64/libc.so.6
#4  0x0000000000784c1b in tracy::Worker::PackPointer(unsigned long) const (this=0x15e00f8, ptr=7929556913241470265) at ../../../server/TracyWorker.cpp:3531
#5  0x0000000000752647 in tracy::Worker::AddCallstackPayload(unsigned long, char const*, unsigned long) (_sz=16, _data=0x1717cf5 "\365\242a\230\266\177", ptr=140422089932928, this=0x15e00f8) at ../../../server/TracyWorker.cpp:3558
#6  tracy::Worker::DispatchProcess(tracy::QueueItem const&, char const*&) (ptr=<optimized out>, ev=..., this=0x15e00f8) at ../../../server/TracyWorker.cpp:2985
#7  tracy::Worker::Exec() (this=0x15e00f8) at ../../../server/TracyWorker.cpp:2666
#8  0x0000000000726baf in operator()() const (__closure=0x16f3698) at ../../../server/TracyWorker.cpp:271
#9  0x00000000007b5930 in std::__invoke_impl<void, tracy::Worker::Worker(char const*, int)::<lambda()> >(std::__invoke_other, struct {...} &&) (__f=...) at /usr/include/c++/10/bits/invoke.h:60
#10 0x00000000007b568d in std::__invoke<tracy::Worker::Worker(char const*, int)::<lambda()> >(struct {...} &&) (__fn=...) at /usr/include/c++/10/bits/invoke.h:95
#11 0x00000000007b53a2 in std::thread::_Invoker<std::tuple<tracy::Worker::Worker(char const*, int)::<lambda()> > >::_M_invoke<0>(std::_Index_tuple<0>) (this=0x16f3698) at /usr/include/c++/10/thread:264
#12 0x00000000007b5216 in std::thread::_Invoker<std::tuple<tracy::Worker::Worker(char const*, int)::<lambda()> > >::operator()(void) (this=0x16f3698) at /usr/include/c++/10/thread:271
#13 0x00000000007b47be in std::thread::_State_impl<std::thread::_Invoker<std::tuple<tracy::Worker::Worker(char const*, int)::<lambda()> > > >::_M_run(void) (this=0x16f3690) at /usr/include/c++/10/thread:215
#14 0x00007ffff67344d4 in std::execute_native_thread_routine(void*) (__p=0x16f3690) at ../../../../../libstdc++-v3/src/c++11/thread.cc:80
#15 0x00007ffff77eaeaa in start_thread () at /lib64/libpthread.so.0
#16 0x00007ffff6435aff in clone () at /lib64/libc.so.6
```
The full coredump can be downloaded [here](https://perret.website/c173594d399dd7e98c/coredump)

**Comment by wolfpld** [comment]
The pointer that is to be packed is invalid: 0x6E'0B'72'00'00'00'35'39 (see 1.1.3 in http://developer.amd.com/wordpress/media/2012/10/24593_APM_v2.pdf; current CPUs implement 48 bit address space).

Can you check if pointers reported by kernel are correct (and maybe, if one of them matches the faulting pointer in the server)? List of pointers in each stack frame can be dumped to a file at `client/TracySysTrace.cpp:745`, from the `trace` array of `uint64_t` values, where the first value is count of pointers (`uint64_t` values) stored at the following addresses.

**Comment by occivink** [comment]
I definitely get different values than the ones dumped from the client, which look more like `0xd35d6938`. None of them match the faulty one reported by the server.

I can also reproduce the issue on a different computer running arch linux, with different hardware as well (intel vs amd cpu).

**Comment by wolfpld** [comment]
Ok, I'll look into this more closely.

**Comment by wolfpld** [comment]
This seems to be a bug in the kernel. No surprise there.

The documentation (https://man7.org/linux/man-pages/man2/perf_event_open.2.html) says:

```
nr, ips[nr]
    If PERF_SAMPLE_CALLCHAIN is enabled, then a 64-bit
    number is included which indicates how many following
    64-bit instruction pointers will follow.  This is the
    current callchain.
```

This is what I get _on the main thread_. The zeroth element is `nr`, with `ips[nr]` following:

```
0. 0x4
1. 0xfffffffffffffe00
2. 0x55da661808e9
3. 0x7f6e8d496152
4. 0x4956410000d5eb3d
```

This is what I get _on another thread_:

```
0. 0x5
1. 0xfffffffffffffe00
2. 0x7f6e8d98d00c
3. 0x55da6616c10b
4. 0x55da66183052
5. 0x7f6e8d9833e9
```

As you can see, on main thread a bullshit pointer is provided.

Looking at some older traces I can see this was always the case (the release build ignores assertions). What's interesting is the similarity in what is being returned as the last pointer:

```
Linux 5.8.1 (run 1): 0x09564100'3519b33d
Linux 5.8.1 (run 2): 0x09564100'3528033d
Linux 5.8.1 (run 3): 0x09564100'00c94b3d - this is after a reboot
Linux 5.8.12:        0x49564100'00d5eb3d
Yours:               0x6e0b7200'00003539
Linux 4.4.222:       0x0f6f6c99'6176816c - android
```

**Comment by trishume** [comment]
`PERF_SAMPLE_CALLCHAIN` relies on the frame pointer to decode the stack, but many things on Linux aren't compiled with frame pointers enabled (`-fomit-frame-pointer` and `-fno-omit-frame-pointer` control this), including often the C standard library. When the kernel tries to sample the call stack of an app which uses the frame pointer as a general purpose register I don't know exactly what happens but that's probably how it ends up with invalid pointers, since there's no way for the kernel to know that the frame pointer is invalid until it fails to follow a frame pointer, and it probably tries to give as many frames as it can in case they're valid.

Basically on Linux if you're using frame pointers to sample the call stack you need to gracefully handle invalid stacks and also you often won't get valid stacks unless the app is compiled with frame pointers.

**Comment by wolfpld** [comment]
A very short test run produced many different invalid pointers:

```
0x1008e8c7890000ff
0x11f88308408b8845
0x15ab8058d480000
0x18858948ffffe0ac
0x18958b48ffffeb66
0x1b8010000014085
0x1c48341f7894c20
0x1e0c90d8d481f7f
0x1f0f66e0fff80148
0x203a746968736c6c
0x24448b48c1894800
0x251830f000008
0x2520757063007461
...
```

**Comment by wolfpld** [comment]
With fc1b03d6 I no longer get assertion on the server.

**Comment by occivink** [comment]
I don't get this assertion anymore, but now I "randomly" get 
```
Tracy-debug: ../../../server/TracyWorker.cpp:5885: void tracy::Worker::ProcessContextSwitch(const tracy::QueueContextSwitch&): Assertion `m_data.externalThreadCompress.DecompressThread( cx.Thread() ) == ev.oldThread' failed.
```
as well as 
```
Tracy-debug: ../../../server/TracyWorker.cpp:3662: tracy::CallstackFrameId tracy::Worker::PackPointer(uint64_t) const: Assertion `( ( ptr & 0x3000000000000000 ) << 2 ) == ( ptr & 0xC000000000000000 )' failed.
```

**Comment by wolfpld** [comment]
The three assertions you have listed all indicate that there are problems with promptness of handling data from kernel ring buffers. I see that you are using debug version of the server (why?), so I'll ask if you're also profiling a debug version of your application?

**Comment by occivink** [comment]
You're right that I was running a debug of build of tracy, that was simply because it's the default when calling `make`. The application I was profiling was compiled in `release` mode, though. In any case, switching tracy to a release build seem not to trigger these assertions.

**Comment by wolfpld** [comment]
Can you provide more details about your hardware and number of threads your application is using? Tracy sets a pretty high priority for kernel buffer handling thread and it still seems to be starved.

Asserts at 5885 and 5910 mean that the context switch capture is missing some data, which is not handled too well right now. You will encounter various problems due to this.

Assert at 5525 will break some things during the real-time capture, but it will self-correct after saving and reloading the trace.

**Comment by occivink** [comment]
>Can you provide more details about your hardware and number of threads your application is using?

Interestingly the application does not use any additional threads, so it might have been caused by other processes. I was still trying to get familiar with tracy at the time so I was not seriously profiling it. I cannot reproduce this when the rest of the system is (mostly) idle. 
The CPU I'm using is an intel skylake 6600k.

I do still get this assertion, but only spuriously after starting the server (~5s), whereas it used to be right away.
```
Tracy-debug: ../../../server/TracyWorker.cpp:3662: tracy::CallstackFrameId tracy::Worker::PackPointer(uint64_t) const: Assertion `( ( ptr & 0x3000000000000000 ) << 2 ) == ( ptr & 0xC000000000000000 )' failed.
```

Thanks for your help.

**Comment by wolfpld** [comment]
Yes, it may be necessary to scan the entire call stack, not only the last frame. Will get back to this when I have some time, but this issue is harmless, aside from triggering an assert and showing an unknown frame in call stack display.

**Comment by wolfpld** [comment]
Yup, there are bad pointers in the middle of stack trace.

```
 0. 0x7f2f98468940
 1. 0xfffffffffffffe00
 2. 0x7f2f9ff8979e
 3. 0x4000000000000000
 4. 0x7ffffc50ee41
```

**Comment by wolfpld** [comment]
Should be fixed by d7550304.

**Comment by emoon** [comment]
Hey,

I'm running on an ARM device with version `0.11.1` and I connect remotely to it from my x64 machine and I occasionally get this assert

```
cy-profiler: tracy/tracy-0.11.1/server/TracyWorker.cpp:3907: tracy::CallstackFrameId tracy::Worker::PackPointer(uint64_t) const: Assertion `( ( ptr & 0x3000000000000000 ) << 2 ) == ( ptr & 0xC000000000000000 )' failed.
```
Callstack
```
* thread #15, name = 'Tracy Worker', stop reason = signal SIGABRT
  * frame #0: 0x00007ffff7877624 libc.so.6`__pthread_kill_implementation(threadid=<unavailable>, signo=6, no_tid=<unavailable>) at pthread_kill.c:44:76
    frame #1: 0x00007ffff781dba0 libc.so.6`__GI_raise(sig=6) at raise.c:26:13
    frame #2: 0x00007ffff7805582 libc.so.6`__GI_abort at abort.c:73:3
    frame #3: 0x00007ffff78054eb libc.so.6`__assert_fail_base(fmt=<unavailable>, assertion=<unavailable>, file=<unavailable>, line=3907, function=<unavailable>) at assert.c:118:3
    frame #4: 0x00005555560bbb96 tracy-profiler`tracy::Worker::PackPointer(unsigned long) const + 98
    frame #5: 0x0000555556055803 tracy-profiler`tracy::Worker::Exec() + 6431
    frame #6: 0x000055555603e072 tracy-profiler`tracy::Worker::Worker(char const*, unsigned short, long)::'lambda'()::operator()() const + 42
    frame #7: 0x0000555556167733 tracy-profiler`void std::__invoke_impl<void, tracy::Worker::Worker(char const*, unsigned short, long)::'lambda'()>(std::__invoke_other, tracy::Worker::Worker(char const*, unsigned short, long)::'lambda'()&&) + 32
    frame #8: 0x0000555556166fa6 tracy-profiler`std::__invoke_result<tracy::Worker::Worker(char const*, unsigned short, long)::'lambda'()>::type std::__invoke<tracy::Worker::Worker(char const*, unsigned short, long)::'lambda'()>(tracy::Worker::Worker(char const*, unsigned short, long)::'lambda'()&&) + 32
    frame #9: 0x0000555556166472 tracy-profiler`void std::thread::_Invoker<std::tuple<tracy::Worker::Worker(char const*, unsigned short, long)::'lambda'()>>::_M_invoke<0ul>(std::_Index_tuple<0ul>) + 40
    frame #10: 0x0000555556165872 tracy-profiler`std::thread::_Invoker<std::tuple<tracy::Worker::Worker(char const*, unsigned short, long)::'lambda'()>>::operator()() + 24
    frame #11: 0x0000555556163698 tracy-profiler`std::thread::_State_impl<std::thread::_Invoker<std::tuple<tracy::Worker::Worker(char const*, unsigned short, long)::'lambda'()>>>::_M_run() + 28
    frame #12: 0x00007ffff7ae1f74 libstdc++.so.6`std::execute_native_thread_routine(__p=0x0000555557b4a9f0) at thread.cc:104:18
    frame #13: 0x00007ffff787570a libc.so.6`start_thread(arg=<unavailable>) at pthread_create.c:448:8
    frame #14: 0x00007ffff78f9aac libc.so.6`__clone3 at clone3.S:78
```

**Comment by wolfpld** [comment]
Tracy assumes pointers are canonical (https://stackoverflow.com/questions/38977755/address-canonical-form-and-pointer-arithmetic). ARM has extensions that may break this assumption (https://source.android.com/docs/security/test/tagged-pointers). There is no workaround, but maybe you can reconfigure your device to make it work.

**Comment by wolfpld** [comment]
It may work, or it may break. I don't remember the exact assumptions. Top bits are used by Tracy to store some metadata, and there needs to be a way to reconstruct the original pointer, which is trivial with the canonical form. It's possible that Tracy may want to compare the reconstructed pointer against a raw received pointer, and these won't match, when they should.

**Comment by emoon** [comment]
I see. Thanks!
